Bug 1878602: Support for stall daemon.

jmencak opened the pull request:
This PR adds support for the stalld program to prevent the _starvation_ of
operating system threads in a Linux system.

Changes:
  - support for `[service]` Tuned plugin
  - support the installation of stalld binary and systemd units on hosts fully-managed by the [Machine Config Operator](https://github.com/openshift/machine-config-operator) (MCO).

Example usage:

```
$ oc label node <node-name> node-role.kubernetes.io/worker-rt=

$ oc create -f- <<EOF
apiVersion: tuned.openshift.io/v1
kind: Tuned
metadata:
  name: openshift-realtime
  namespace: openshift-cluster-node-tuning-operator
spec:
  profile:
  - data: |
      [main]
      summary=Custom OpenShift realtime profile
      include=openshift-node,realtime
      [service]
      service.stalld=start,enable
    name: openshift-realtime

  recommend:
  - machineConfigLabels:
      machineconfiguration.openshift.io/role: "worker-rt"
    priority: 30
    profile: openshift-realtime
EOF

$ oc create -f examples/realtime-mcp.yaml
```

A host-supplied configuration file can be used to override the `stalld` systemd unit.  This file can then be referred to by prefixing the absolute path to the file on the host by the `/host` prefix.  For example:

```
      [service]
      service.stalld=start,enable,file:/host/etc/tuned/openshift-realtime/stalld.conf
```

Note that the file needs to be created on the host by other means, for example, by using a MachineConfig. One of the uses of the drop-in file is to supply environment variables to the stalld unit.  For example, the drop-in file may contain:
```
[Service]
Environment=CLIST=
Environment=AGGR=
Environment=BP="-p 1000000000"
Environment=BR="-r 10000"
Environment=BD="-d 3"
Environment=THRESH="-t 60"
Environment=LOGGING="--log_syslog --log_kmsg"
Environment=FG=--foreground
Environment=PF="--pidfile /run/stalld.pid"
```

jmencak commented:
/cc @MarSik 
/cc @yanirq

kpouget reviewed (commented):
I didn't test the code, but what I read/understood in the patch seem sensible and consistent, thanks!
I only made a minor style question.

I didn't read in detail the Python code (the service plugin patch), is it new code or is it coming from elsewhere?

jmencak commented on pkg/operator/controller.go:
```diff
@@ -561,48 +564,84 @@ func (c *Controller) syncProfile(tuned *tunedv1.Tuned, nodeName string) error {
-	if util.StringSlicesEqual(mc.Spec.KernelArguments, kernelArguments) {
+	mcNew := newMachineConfig(name, annotations, labels, kernelArguments, ignFiles, ignUnits)
+
+	ignEqual, err := ignEqual(mc, mcNew)
```
Yeah, good point.  And even if it was safe, this can probably be very confusing for people to read.  Let me change the variable or function name.

jmencak commented:
The "Operand cleanup, using long option for stalld." got rid of the limitation in the description.  Removing.  This is a Friday night commit -- it seems to work but more testing and cleanup needed.

yanirq commented on examples/realtime-mc.yaml:
```diff
@@ -16,6 +16,8 @@ spec:
       [bootloader]
       cmdline_ocp_realtime=+systemd.cpu_affinity=${not_isolated_cores_expanded}
+      #[service]
+      #service.stalld=start,enable
```
passing variables will be here ? (if not then a reference in the comment will suffice)

jmencak replied:
Yes, I'm planning to document this.  A "pointer" to an environment file will be passed here.

yanirq commented:
Will tests be added in this PR ? (should be)

jmencak commented:
> Will tests be added in this PR ? (should be)

As time allows one all issues are resolved.  I still see two issues that need to be addressed.

jmencak commented:
/test e2e-aws-upgrade

yanirq commented:
I would add the PR's description to NTO's readme section (or open a new sub folder doc)

jmencak commented:
@yanirq cannot reproduce the "degraded" MCP state issues I was having last week after several successful tests.  There were some changes in MCO which hopefully fixed those.  I believe I've addressed all your concerns?  Could you please have another look through the code and indicate whether there's something missing?  Thanks!

yanirq commented:
/retest

yanirq commented:
/lgtm
/hold
looks good overall (unless im missing something) if you feel comfortable release the hold 
I do think we need tests embedded (they usually come with the PR) but because of urgency we can merge then in a very soon to follow PR

jmencak commented:
> /lgtm
> /hold
> looks good overall (unless im missing something) if you feel comfortable release the hold
> I do think we need tests embedded (they usually come with the PR) but because of urgency we can merge then in a very soon to follow PR

Thanks @yanirq . Which tests do you have in mind?  There is an embedded e2e tests for this new functionality.

jmencak commented:
@kpouget / @dagrayvid  would you be able to provide a feedback for this PR?  This is close to being complete.  Thank you!

jmencak commented:
/bugzilla refresh

kpouget commented on pkg/tuned/host_payload.go:
```diff
@@ -0,0 +1,74 @@
+[Service]
+Environment=CLIST=
+Environment=AGGR=
+Environment=BP="-p 1000000000"
+Environment=BR="-r 20000"
+Environment=BD="-d 3"
```
really just a suggestion: long options could be easier to understand, `boost_period` instead of `BP` or `p`

```suggestion
Environment=BP="--boost_period 1000000000"
Environment=BR="--boost_runtime 20000"
Environment=BD="--boost_duration 3"
```

jmencak replied:
Agreed, however, I tried to diverge from the upstream as little as possible (https://github.com/bristot/stalld/blob/master/redhat/stalld.conf).  Honestly not sure which is better.  Diverge or have long options.

jmencak replied:
Added comments in the unit to match the upstream `stalld.conf`.

jmencak commented:
Reiterating
/hold
from Yanir as the commit messages need to be squashed before merge.  We also need another review from PSAP team.

kpouget commented on pkg/tuned/tuned.go:
```diff
@@ -370,11 +380,17 @@ func (c *Controller) tunedRun() {
 		}
 	}()
 
 	err = c.tunedCmd.Start()
+	c.daemon.reloading = true
```
I'm confused by the lines above, I'm afraid that there might be a race condition, but I have 0 prior knowledge about Goroutines, so take this comment with caution:
* `c.tunedCmd.Start()` unblocks `scanner.Text()`, and the goroutine checks `c.daemon.reloading`
* the main routine sets `	c.daemon.reloading = true`

```suggestion
	c.daemon.reloading = true
	err = c.tunedCmd.Start()
```

jmencak replied:
This is a great catch, thank you!  Will fix.

kpouget reviewed (commented):
looks good to me, except a possible race condition. Please have a look and let me know if you agree, or if I missed something in the synchronization!

jmencak commented on assets/tuned/patches/070-service-plugin.diff:
```diff
@@ -0,0 +1,354 @@
+ DEF_CGROUP_MODE = 0o770
+ 
++# service plugin configuration
++SERVICE_SYSTEMD_CFG_PATH="/etc/systemd/system/%s.service.d"
```
Thanks.  This is an official upstream tuned patch.  Cannot change it here, but can raise this in upstream tuned.

jmencak replied:
Created a PR to get this fixed upstream.

https://github.com/redhat-performance/tuned/pull/295

jmencak commented:
/retest

jmencak commented:
/hold cancel
I believe I've addressed reviewer's comments.  Commits squashed.  If this still looks good, @yanirq feel free to lgtm.

yanirq commented:
/lgtm
